File: src/APIBundle/Entity/Group.php
<?php

namespace APIBundle\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\ORM\Mapping as ORM;

/**
 * @ORM\Entity()
 * @ORM\Table(name="groups")
 */
class Group
{
    /**
     * @ORM\Id
     * @ORM\Column(type="string")
     * @ORM\GeneratedValue(strategy="UUID")
     */
    protected $group_id;

    /**
     * @ORM\Column(type="string")
     */
    protected $group_name;

    /**
     * @ORM\ManyToMany(targetEntity="User", inversedBy="group_users")
     * @ORM\JoinTable(name="users_groups",
     *      joinColumns={@ORM\JoinColumn(name="group_id", referencedColumnName="group_id", onDelete="CASCADE")},
     *      inverseJoinColumns={@ORM\JoinColumn(name="user_id", referencedColumnName="user_id", onDelete="CASCADE")}
     *      )
     * @var User[]
     */
    protected $group_users;

    /**
     * @ORM\OneToMany(targetEntity="Grant", mappedBy="grant_group")
     * @var Grant[]
     */
    protected $group_grants;

    /**
     * @ORM\ManyToOne(targetEntity="User")
     * @ORM\JoinColumn(name="group_owner", referencedColumnName="user_id")
     */
    protected $group_owner;

    protected $group_grant_in_exercise;

    public function __construct()
    {
        $this->group_users = new ArrayCollection();
        $this->group_grants = new ArrayCollection();
    }

    public function getGroupId()
    {
        return $this->group_id;
    }

    public function setGroupId($id)
    {
        $this->group_id = $id;
        return $this;
    }

    public function getGroupName()
    {
        return $this->group_name;
    }

    public function setGroupName($name)
    {
        $this->group_name = $name;
        return $this;
    }

    public function getGroupUsers()
    {
        return $this->group_users;
    }

    public function setGroupUsers($users)
    {
        $this->group_users = $users;
        return $this;
    }

    public function getGroupGrants()
    {
        return $this->group_grants;
    }

    public function setGroupGrants($grants)
    {
        $this->group_grants = $grants;
        return $this;
    }

    public function getGroupOwner()
    {
        return $this->group_owner;
    }

    public function setGroupOwner($owner)
    {
        $this->group_owner = $owner;
        return $this;
    }

    public function getGroupGrantInExercise()
    {
        return $this->group_grant_in_exercise;
    }

    public function setGroupGrantInExercise($grant)
    {
        $this->group_grant_in_exercise = $grant;
        return $this;
    }
}

File: openbas-model/src/main/java/io/openbas/database/model/Group.java
package io.openbas.database.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import io.openbas.annotation.Queryable;
import io.openbas.database.audit.ModelBaseListener;
import io.openbas.helper.MultiIdListDeserializer;
import io.openbas.helper.MultiModelDeserializer;
import io.swagger.v3.oas.annotations.media.ArraySchema;
import io.swagger.v3.oas.annotations.media.Schema;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotBlank;
import java.util.*;
import java.util.stream.Collectors;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.Fetch;
import org.hibernate.annotations.FetchMode;
import org.hibernate.annotations.UuidGenerator;

@Setter
@Getter
@Entity
@Table(name = "groups")
@EntityListeners(ModelBaseListener.class)
public class Group implements Base {

  @Id
  @Column(name = "group_id")
  @GeneratedValue(generator = "UUID")
  @UuidGenerator
  @JsonProperty("group_id")
  @NotBlank
  private String id;

  @Queryable(sortable = true)
  @Column(name = "group_name")
  @JsonProperty("group_name")
  @NotBlank
  private String name;

  @Column(name = "group_description")
  @JsonProperty("group_description")
  private String description;

  @Queryable(sortable = true)
  @Column(name = "group_default_user_assign")
  @JsonProperty("group_default_user_assign")
  private boolean defaultUserAssignation;

  @ElementCollection(fetch = FetchType.EAGER)
  @CollectionTable(name = "groups_default_grants", joinColumns = @JoinColumn(name = "group_id"))
  @JsonProperty("group_default_grants")
  @JsonInclude(JsonInclude.Include.NON_NULL)
  private Set<DefaultGrant> defaultGrants = new HashSet<>();

  @OneToMany(
      mappedBy = "group",
      fetch = FetchType.EAGER,
      cascade = CascadeType.ALL,
      orphanRemoval = true)
  @JsonProperty("group_grants")
  @JsonSerialize(using = MultiModelDeserializer.class)
  @Fetch(value = FetchMode.SUBSELECT)
  private List<Grant> grants = new ArrayList<>();

  @ArraySchema(schema = @Schema(type = "string"))
  @ManyToMany(fetch = FetchType.EAGER)
  @JoinTable(
      name = "users_groups",
      joinColumns = @JoinColumn(name = "group_id"),
      inverseJoinColumns = @JoinColumn(name = "user_id"))
  @JsonSerialize(using = MultiIdListDeserializer.class)
  @JsonProperty("group_users")
  @Fetch(value = FetchMode.SUBSELECT)
  private List<User> users = new ArrayList<>();

  @ArraySchema(schema = @Schema(type = "string"))
  @ManyToMany(fetch = FetchType.LAZY)
  @JoinTable(
      name = "groups_organizations",
      joinColumns = @JoinColumn(name = "group_id"),
      inverseJoinColumns = @JoinColumn(name = "organization_id"))
  @JsonSerialize(using = MultiIdListDeserializer.class)
  @JsonProperty("group_organizations")
  private List<Organization> organizations = new ArrayList<>();

  @ArraySchema(schema = @Schema(type = "string"))
  @ManyToMany(fetch = FetchType.EAGER)
  @JoinTable(
      name = "groups_roles",
      joinColumns = @JoinColumn(name = "group_id"),
      inverseJoinColumns = @JoinColumn(name = "role_id"))
  @JsonSerialize(using = MultiIdListDeserializer.class)
  @JsonProperty("group_roles")
  private List<Role> roles = new ArrayList<>();

  @Getter(onMethod_ = @JsonIgnore)
  @Transient
  private final ResourceType resourceType = ResourceType.USER_GROUP;

  // region transient
  // Convert to Map for easier access
  @Transient
  @JsonIgnore
  public Map<Grant.GRANT_RESOURCE_TYPE, List<Grant.GRANT_TYPE>> getDefaultGrantsMap() {
    return defaultGrants.stream()
        .collect(
            Collectors.groupingBy(
                DefaultGrant::getGrantResourceType,
                Collectors.mapping(DefaultGrant::getGrantType, Collectors.toList())));
  }

  @JsonIgnore
  @Transient
  public List<Grant.GRANT_TYPE> getSimulationsDefaultGrants() {
    return defaultGrants.stream()
        .filter(dg -> dg.getGrantResourceType() == Grant.GRANT_RESOURCE_TYPE.SIMULATION)
        .map(DefaultGrant::getGrantType)
        .collect(Collectors.toList());
  }

  @JsonIgnore
  @Transient
  public List<Grant.GRANT_TYPE> getScenariosDefaultGrants() {
    return defaultGrants.stream()
        .filter(dg -> dg.getGrantResourceType() == Grant.GRANT_RESOURCE_TYPE.SCENARIO)
        .map(DefaultGrant::getGrantType)
        .collect(Collectors.toList());
  }

  // endregion

  @Override
  public boolean isUserHasAccess(User user) {
    return users.contains(user);
  }

  @Override
  public boolean equals(Object o) {
    if (this == o) return true;
    if (o == null || !Base.class.isAssignableFrom(o.getClass())) return false;
    Base base = (Base) o;
    return id.equals(base.getId());
  }

  @Override
  public int hashCode() {
    return Objects.hash(id);
  }
}
